kvpb: add NoReplicasMoved to AdminScatterResponse

wenyihu6 opened the pull request:
**kvpb: remove outdated comment for AdminScatterRequest**

Previously, AdminScatterResponse included error details when range moves failed
(https://github.com/cockroachdb/cockroach/blob/225f782748bd13656a3a5ce99f3a5369be4a76a3/pkg/roachpb/api.proto#L849-L853).
These details were removed since
https://github.com/cockroachdb/cockroach/commit/dbd90cf0c331c1f109985a22f96937e75e46aa52.
This commit removes the now-outdated comment referring to them.

Epic: none
Release note: none

---

**kvpb: add NoReplicasMoved to AdminScatterResponse**

Previously, clients couldn't determine if the scatter operation from
AdminScatterRequest succeeded, as failures to move ranges would not result in a
client error being returned. Since AdminScatter is used in multiple places, we
chose to preserve the existing best-effort behavior and avoid surfacing these
errors to the client. But to improve visibility for clients, this commit
introduces a new no_replicas_moved field (defaulting to false for mixed-version
compatibility). This field indicates that no replicas were moved, based on
comparing the state before and after the scatter operation, suggesting the
operation may warrant further attention or a retry.

Note: the client side doesn't currently read this field, so this change doesn't
affect behavior yet.

Resolves: #144560
Resolves: #144548
Resolves: #144571
Release note: none

---

**backup: add warning when no replica moved from AdminScatter**

Previously, we introduced a new field NoReplicasMoved to
AdminScatterResponse. This is to provide more visibility into
the scatter operation to indicate whether AdminScatter actually
successfully moved replicas.

This commit just logs a warning if NoReplicasMoved is true
from the DR side. Note that retrying from the client side is
not implemented, so no behaviour changes exist yet. This remains
as an ongoing issue tracked in #144856.

Informs: #144856
Part of: #144560
Release note: none

wenyihu6 committed:
kvpb: remove outdated comment for AdminScatterRequest

Previously, AdminScatterResponse included error details when range moves failed
(https://github.com/cockroachdb/cockroach/blob/225f782748bd13656a3a5ce99f3a5369be4a76a3/pkg/roachpb/api.proto#L849-L853).
These details were removed since
https://github.com/cockroachdb/cockroach/commit/dbd90cf0c331c1f109985a22f96937e75e46aa52.
This commit removes the now-outdated comment referring to them.

Epic: none
Release note: none

wenyihu6 committed:
kvpb: add NoReplicasMoved to AdminScatterResponse

Previously, clients couldn't determine if the scatter operation from
AdminScatterRequest succeeded, as failures to move ranges would not result in a
client error being returned. Since AdminScatter is used in multiple places, we
chose to preserve the existing best-effort behavior and avoid surfacing these
errors to the client. But to improve visibility for clients, this commit
introduces a new no_replicas_moved field (defaulting to false for mixed-version
compatibility). This field indicates that no replicas were moved, based on
comparing the state before and after the scatter operation, suggesting the
operation may warrant further attention or a retry.

Note: the client side doesn't currently read this field, so this change doesn't
affect behavior yet.

Resolves: #144560
Resolves: #144548
Resolves: #144571
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144845)
<!-- Reviewable:end -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo nits

> Reviewed 1 of 1 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-:-OOSlZxf2ZFV21EpmPjR:b-1p9zor)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @msbutler)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSkHHgEuOj3WxgXeAF:-OOSkHHgEuOj3WxgXeAG:b7wfadi):*
:detective: 
___
*[`-- commits` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSlEgP8CuR1oS3lg8C:-OOSlEgP8CuR1oS3lg8D:b-aa8mv8):*
nit: we're reading this on the client, just not acting upon it for now.
___
*[`pkg/kv/kvserver/replica_command.go` line 4269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSlLu67pgDKmrxizah:-OOSlLu67pgDKmrxizai:bk1l3bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/kv/kvserver/replica_command.go#L4269)):*
> ```Go
> 		// adequate.
> 		ReplicasScatteredBytes: stats.Total() * int64(numReplicasMoved),
> 		// Note, this needs to be a separate field, since we can't reliably infer
> ```

nit: s/Note/NB:/

Separately, I wonder if this comment is better placed at the proto definition instead (could duplicate it too).
___
*[`pkg/kv/kvserver/replica_command.go` line 4272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSkS5EE568KT_D1Bag:-OOSkS5F-YSF0kXKrG9V:b-vhccs0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/kv/kvserver/replica_command.go#L4272)):*
> ```Go
> 		// "no replica moved" from ReplicasScatteredBytes == 0.Empty ranges have
> 		// stats.Total() == 0 and ReplicasScatteredBytes will be 0 regardless of
> 		// whether no replica was moved.
> ```

nit: "whether any replicas were moved"
___
*[`pkg/backup/generative_split_and_scatter_processor.go` line 215 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSktovAfYX67445A1m:-OOSktovAfYX67445A1n:b-qspdha) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/backup/generative_split_and_scatter_processor.go#L215)):*
> ```Go
> 			log.Errorf(ctx, "failed to scatter span [%s,%s): %+v",
> 				newScatterKey, newScatterKey.Next(), pErr.GoError())
> 		} else {
> ```

nit: changes to this file could have been  their own commit.
___
*[`pkg/backup/generative_split_and_scatter_processor.go` line 229 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSl5Rz70jFjaw0tRT7:-OOSl5Rz70jFjaw0tRT8:b-jvy2es) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/backup/generative_split_and_scatter_processor.go#L229)):*
> ```Go
> 		log.Fatalf(ctx, "expected AdminScatterResponse, got %T", res)
> 	}
> 	if scatteredResp.NoReplicasMoved {
> ```

Should we add a TODO to retry if this is the case? 
___
*[`pkg/kv/kvpb/api.go` line 629 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSkebTBZzsybhRa7wh:-OOSkebTBZzsybhRa7wi:b5fsk19) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/kv/kvpb/api.go#L629)):*
> ```Go
> 		r.MVCCStats.Add(otherR.MVCCStats)
> 		r.ReplicasScatteredBytes += otherR.ReplicasScatteredBytes
> 		r.NoReplicasMoved = r.NoReplicasMoved && otherR.NoReplicasMoved
> ```

I don't even know what it means to combine two scatter responses, but should this be an or instead of an and? 


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
backup: add warning when no replica moved from AdminScatter

Previously, we introduced a new field NoReplicasMoved to
AdminScatterResponse. This is to provide more visibility into
the scatter operation to indicate whether AdminScatter actually
successfully moved replicas.

This commit just logs a warning if NoReplicasMoved is true
from the DR side. Note that retrying from the client side is
not implemented, so no behaviour changes exist yet. This remains
as an ongoing issue tracked in #144856.

Informs: #144856
Part of: #144560
Release note: none

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @msbutler)

___
*[`-- commits` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSlEgP8CuR1oS3lg8C:-OOT2_SYC3GxuCKdjI7b:b-c0r8tb):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: we're reading this on the client, just not acting upon it for now.
</blockquote></details>

I added a new commit for the dr side of changes as suggested by https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSktovAfYX67445A1m. This comment seems accurate now as a result.
___
*[`pkg/backup/generative_split_and_scatter_processor.go` line 215 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSktovAfYX67445A1m:-OOT2loO0RAATxyLwXQa:b-32muyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/backup/generative_split_and_scatter_processor.go#L215)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: changes to this file could have been  their own commit.
</blockquote></details>

Good call, splitted into another commit in the end. 
___
*[`pkg/backup/generative_split_and_scatter_processor.go` line 229 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSl5Rz70jFjaw0tRT7:-OOT3xdu7iNURmmI33hz:bdvps80) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/backup/generative_split_and_scatter_processor.go#L229)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we add a TODO to retry if this is the case?
</blockquote></details>

Filed #144856. 
___
*[`pkg/kv/kvpb/api.go` line 629 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSkebTBZzsybhRa7wh:-OOT4wys8uoK6I3gFtNF:b-oo1kuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/kv/kvpb/api.go#L629)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't even know what it means to combine two scatter responses, but should this be an or instead of an and?
</blockquote></details>

+1, I found this really confusing - especially since we're combining `RangeInfos` here. I'm wondering if we should just make it non-combinable. I was going to remove https://github.com/cockroachdb/cockroach/blob/8f00544fbc87b428671f90f348dadfdd519d8883/pkg/kv/kvpb/api.go#L619 entirely. But I’m not sure what this would entail - when does response combining happens and if changing this could cause a perf regression. 
___
*[`pkg/kv/kvserver/replica_command.go` line 4269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSlLu67pgDKmrxizah:-OOT5I7k80UNFw8eZsNV:bzb91h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/kv/kvserver/replica_command.go#L4269)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/Note/NB:/

Separately, I wonder if this comment is better placed at the proto definition instead (could duplicate it too).
</blockquote></details>

Done. Duplicated it to proto definition. 
___
*[`pkg/kv/kvserver/replica_command.go` line 4272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSkS5EE568KT_D1Bag:-OOT5Lbe59hrayHhqMtn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/kv/kvserver/replica_command.go#L4272)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: "whether any replicas were moved"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @msbutler)

___
*[`pkg/kv/kvpb/api.go` line 629 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144845#-OOSkebTBZzsybhRa7wh:-OOTKR92EGPsaTSeT67y:b5t9qqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9925305858504a794118556f7780f1fc8d76e1/pkg/kv/kvpb/api.go#L629)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

+1, I found this really confusing - especially since we're combining `RangeInfos` here. I'm wondering if we should just make it non-combinable. I was going to remove https://github.com/cockroachdb/cockroach/blob/8f00544fbc87b428671f90f348dadfdd519d8883/pkg/kv/kvpb/api.go#L619 entirely. But I’m not sure what this would entail - when does response combining happens and if changing this could cause a perf regression.
</blockquote></details>

Filed https://github.com/cockroachdb/cockroach/issues/144864 to track the follow up work on this. Going to merge this patch as is. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=arulajmani
